core/src/test/scala/org/apache/spark/CacheRecoveryManagerSuite.scala

```diff
@@ -0,0 +1,200 @@
+    val blocks = Seq(RDDBlockId(1, 1), RDDBlockId(2, 1))
+    val bmme = FakeBMM(1, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+    when(eam.killExecutors(Seq("1"))).thenReturn(Seq("1"))
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+
+    eventually {
+      verify(eam).killExecutors(Seq("1"))
+      bmme.replicated.get("1").get shouldBe 2
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("CacheRecoveryManager will kill executor if it takes too long to replicate") {
+    val conf = new SparkConf().set(DYN_ALLOCATION_CACHE_RECOVERY_TIMEOUT.key, "1s")
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Set(RDDBlockId(1, 1), RDDBlockId(2, 1), RDDBlockId(3, 1), RDDBlockId(4, 1))
+    val bmme = FakeBMM(600, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+
+    eventually(timeout(Span(1010, Millis)), interval(Span(500, Millis))) {
+      verify(eam, times(1)).killExecutors(Seq("1"))
+      bmme.replicated.get("1").get shouldBe 1
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("shutdown timer will get cancelled if replication finishes") {
+    val conf = new SparkConf().set(DYN_ALLOCATION_CACHE_RECOVERY_TIMEOUT.key, "1s")
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Set(RDDBlockId(1, 1))
+    val bmme = FakeBMM(1, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+    val minimumTime = System.currentTimeMillis() + 1000
+
+    eventually(timeout(Span(1500, Millis)), interval(Span(500, Millis))) {
+      // should be killed once not twice
+      verify(eam, times(1)).killExecutors(Seq("1"))
+      // wait at least a second to be sure we don't kill executor twice
+      System.currentTimeMillis() should be > minimumTime
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("Blocks won't replicate if we are running out of space") {
+    val conf = new SparkConf()
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Seq(RDDBlockId(1, 1), RDDBlockId(1, 1), RDDBlockId(1, 1), RDDBlockId(1, 1))
+    val memStatus = Map(BlockManagerId("1", "host", 12, None) -> ((2L, 1L)),
+      BlockManagerId("2", "host", 12, None) -> ((3L, 1L)),
+      BlockManagerId("3", "host", 12, None) -> ((4L, 1L)),
+      BlockManagerId("4", "host", 12, None) -> ((4L, 4L)))
+    val bmme = FakeBMM(1, blocks.iterator, memStatus)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1", "2", "3"))
+    eventually {
```

Comment:
Should probably have an explicit timeout here somewhere? Or maybe check the futures returned by the call above instead.